callstack/linaria
# feat(atomic): add support for atomic using styled API

## Motivation



This continues upon the excellent work by @jpnelson on adding first class (opt-in) Atomic CSS support onto Linaria. The motivation (and community support) can be seen on the original issue #225

## Summary



This PR attempts to use the same patterns in #867 :
- adds a `styled` export on `@linaria/atomic` which simply re-exports the one in `@linaria/react`, but can be used to make `atomic-styled` opt-in
- when using `import { styled } from '@linaria/atomic'`, all the CSS rules within are extracted as atomic classes and re-used across different components
- all the '_static_' CSS (either plain static or interpolations that can be evaluated in build time) are extracted as atomic classes
- dynamic interpolations are converted to unique CSS custom properties based on the exact source code, and then atomized, meaning that different components using the exact same dynamic interpolation will end up re-using the same atomic class.
- using `const Component = styled(AnotherComponent)` should still work as expected (and it should be compatible with `@linaria/react` too). There might be certain edge cases with property priorities (wrapping components currently boosts atomic property priorities by 1, which should be enough in most cases to override AnotherComponent original atom classes).
- as introduced by #973 , you can also rename the local import, such as `import { styled as atomicStyled } from '@linaria/atomic'` and it should still work as expected.

### Examples
Input

```js
import { styled } from '@linaria/atomic';

const Component = styled.div`
  color: blue;
  margin: ${100 / 2}px;
  background-color: ${props => props.color};
`;
```


Output

```js
import { styled } from '@linaria/atomic';
const Component = /*#__PURE__*/styled(\\"div\\")({
  name: \\"Component\\",
  vars: {
    \\"1p69eoh\\": [props => props.color]
  },
  class: \\"atm_7l_13q2bts atm_e2_12xxubj atm_gi_12am3vd atm_2d_15g95l5 c17gu1mi\\",
  atomic: true
});

// CSS Output
/*
CSS:
.atm_7l_13q2bts{color:blue;}
.atm_e2_12xxubj{height:100px;}
.atm_gi_12am3vd{margin:50px;}
.atm_2d_15g95l5.atm_2d_15g95l5{background-color:var(--1p69eoh);}
*/
```

## Test plan



Several tests were added with the behaviour described above, and all existing tests were also passing correctly locally (EDIT: not sure why CI v10 can't resolve the internal dependency, locally seems to work fine and the other CI started passing too). Please check the snapshot generated to confirm the functionality works as you'd expect.

I've also applied this changes on the `/website` and it looks like there's no visual difference when replacing
```diff
- import { styled } from '@linaria/react'
+ import { styled } from '@linaria/atomic'
```

**See the following video**
https://user-images.githubusercontent.com/8507996/163470135-77dccb6d-82c2-49fc-b161-ae9de9435326.mp4
